scheduler/scheduler-node/src/main/java/org/ow2/proactive/scheduler/signal/SignalApiImpl.java

```diff
@@ -0,0 +1,176 @@
+ * or a different license than the AGPL.
+ */
+package org.ow2.proactive.scheduler.signal;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.log4j.Logger;
+import org.awaitility.Awaitility;
+import org.ow2.proactive.scheduler.common.task.TaskId;
+import org.ow2.proactive.scheduler.core.properties.PASchedulerProperties;
+import org.ow2.proactive.scheduler.synchronization.InvalidChannelException;
+import org.ow2.proactive.scheduler.synchronization.SynchronizationInternal;
+import org.ow2.proactive.scheduler.synchronization.SynchronizationWrapper;
+
+
+/**
+ * SignalApiImpl acts as a high level signal service that uses the synchronization API.
+ * It enables workflow tasks handle signals in an easy manner.
+ *
+ * It implements the methods defined in the interface @see org.ow2.proactive.scheduler.signal.SignalApi.
+ *
+ * @author ActiveEon Team
+ * @since 24/11/2020
+ */
+public class SignalApiImpl implements SignalApi {
+
+    private static final Logger logger = Logger.getLogger(SignalApiImpl.class);
+
+    private static final String LOG_WARNING_CONSTANT = " of the job ";
+
+    private static final String SIGNALS_CHANNEL = PASchedulerProperties.SCHEDULER_SIGNALS_CHANNEL.getValueAsString();
+
+    protected static final String READY_PREFIX = "ready_";
+
+    private SynchronizationWrapper synchronization;
+
+    private String jobId;
+
+    public SignalApiImpl(String originator, TaskId taskId, SynchronizationInternal synchronizationInternal) {
+        jobId = taskId.getJobId().value();
+        synchronization = new SynchronizationWrapper(originator, taskId, synchronizationInternal);
+        try {
+
+            synchronization.createChannelIfAbsent(SIGNALS_CHANNEL, true);
+
+            if (!synchronization.containsKey(SIGNALS_CHANNEL, jobId)) {
+                synchronization.put(SIGNALS_CHANNEL, jobId, new ArrayList<>());
+            }
+        } catch (IOException | InvalidChannelException e) {
+            logger.warn("Could not instantiate Signal API for the job " + jobId, e);
+        }
+    }
+
+    @Override
+    public boolean readyForSignal(String signalName) throws InvalidChannelException {
+        if (isReceived(signalName)) {
+            removeSignal(signalName);
+        }
+        return addSignal(READY_PREFIX + signalName);
+    }
+
+    @Override
+    public boolean isReceived(String signalName) throws InvalidChannelException {
+        List<String> signals = getJobSignals();
+        if (!signals.isEmpty()) {
+            return signals.contains(signalName);
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public boolean addSignal(String signalName) {
+        try {
+            List<String> jobSignals = getJobSignals();
+            jobSignals.add(signalName);
+            synchronization.put(SIGNALS_CHANNEL, jobId, (Serializable) jobSignals);
```

Comment:
it could be made atomic, in one line.

synchronization.compute(SIGNALS_CHANNEL, jobId, "{k, x -> x.add('" + signalName "'}");